[MRG] RF: Accept bids_basename in read_raw_bids(), get_matched_empty_room(), get_head_mri_trans(), and drop their bids_fname param

hoechenberger opened the pull request:
<strike>And drop `bids_fname` in favor of `bids_basename`. The filename is automatically populated based on `bids_basename`, `bids_root` and `kind`.

**This is still WIP, as it's already working, but not all tests have been adjusted yet, so CI will fail.**</strike>

See https://github.com/mne-tools/mne-bids/pull/410#issuecomment-629804329

Closes #408.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [x] All comments resolved
- [x] This is not your own PR
- [x] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [x] PR description includes phrase "closes <#issue-number>"
- [x] Commit history does not contain any merge commits

agramfort commented on mne_bids/read.py:
```diff
@@ -328,7 +328,7 @@ def _handle_channels_reading(channels_fname, bids_fname, raw):
 
 
-def read_raw_bids(bids_fname, bids_root, extra_params=None,
+def read_raw_bids(bids_basename, bids_root, extra_params=None, kind=None,
```
@sappelhoff @jasmainak wdyt? breaking backward compat is ok?

if not we need a proper deprecation cycle

sappelhoff replied:
from my side ok to break -> as far as I can see it, most people who use this are active contributors anyway (at least currently).

jasmainak replied:
yeah okay with breaking backward compatibility. I want to know who the users are ... :)

jasmainak commented on mne_bids/read.py:
```diff
@@ -337,29 +337,42 @@ def read_raw_bids(bids_fname, bids_root, extra_params=None,
+        If no data file in a supported format can be located.
 
+    ValueError
+        If the specified ``kind`` cannot be found in the dataset.
```
it's an interesting approach? What other repositories do this? Maybe we should fix this consistently everywhere in a separate PR if you want to do it.

hoechenberger replied:
What _exactly_ do you mean? Documentation of the errors raised?

jasmainak replied:
yep!

jasmainak commented on mne_bids/read.py:
```diff
@@ -369,18 +382,58 @@ def read_raw_bids(bids_fname, bids_root, extra_params=None,
+    valid_exts = ('.fif', '.ds', '.vhdr', '.edf', '.bdf', '.set', '.sqd',
+                  '.con', '.pdf')
+
+    matching_paths = glob.glob(op.join(kind_dir,
+                                       f'{bids_basename}_{kind}' + '.*'))
+    matching_paths = [p for p in matching_paths
+                      if op.splitext(p)[1] in valid_exts]
+
+    if not matching_paths:
+        msg = ('Could not locate a data file of a supported format. This is '
+               'likely a problem with your BIDS dataset. Please run the BIDS '
+               'validator on your data.')
+        raise RuntimeError(msg)
+    # FIXME This will break e.g. with FIFF data split across multiple files
+    if len(matching_paths) > 1:
+        msg = ('Found more than one matching data file for the requested '
+               'recording. Cannot proceed due to the ambiguity. This is '
+               'likely a problem with your BIDS dataset. Please run the BIDS '
+               'validator on your data.')
+        raise RuntimeError(msg)
+
+    matching_path = matching_paths[0]
```
I think this should go in a separate function once you've figured out what makes it work. There are also [some helper functions already in there](https://github.com/mne-tools/mne-bids/blob/25f5b433fdf070ee0942276e8b271523e29bf819/mne_bids/utils.py#L575), so you might want to check if you can reduce duplication.

hoechenberger replied:
I looked and I think for this specific case, the existing function(s) unfortunately don't provide much help

hoechenberger commented:
Wow. Now this blew up. Sorry about that. There were many interconnected things I had to address at once… Just pushed to see what CI thinks. Will explain what I did and why at a later time.

agramfort commented:
tell us when to look

jasmainak commented on mne_bids/tests/test_write.py:
```diff
@@ -226,7 +226,8 @@ def test_fif(_bids_validate):
 
-    raw2 = read_raw_bids(bids_basename + '_eeg.vhdr', bids_root)
+    raw2 = read_raw_bids(bids_basename=bids_basename, bids_root=bids_root,
+                         kind='eeg')
```
do you test `kind=None` anywhere?

hoechenberger replied:
Solved in e6b821b

jasmainak commented on mne_bids/read.py:
```diff
@@ -328,7 +328,117 @@ def _handle_channels_reading(channels_fname, bids_fname, raw):
+                               bids_root=bids_root, sub=sub, ses=ses)
+    if not kinds:
+        raise ValueError('No data found.')
+    elif len(kinds) >= 2:
```
this is not good. What if you have `beh` and `meg`? You should read only `meg`

jasmainak replied:
maybe you should check if, amongst `meg`, `eeg`, and `ieeg`, there are two. I think this is a case you will rarely hit but good to have it.

hoechenberger replied:
Addressed in f51a3f8

jasmainak commented on mne_bids/read.py:
```diff
@@ -328,7 +328,117 @@ def _handle_channels_reading(channels_fname, bids_fname, raw):
+    return kind
+
+
+def _get_bids_fname_from_filesystem(*, bids_basename, bids_root, sub, ses,
```
what is this * ? :) never seen it before

hoechenberger replied:
Keyword-only arguments :)

hoechenberger replied:
Okay with you? I like that because it makes refactoring easier and is more explicit. Also prefer this in parts of public APIs – but that's something for a future discussion :) 

https://www.python.org/dev/peps/pep-3102/

sappelhoff replied:
example for it being used in a public API: https://github.com/xdf-modules/xdf-python/pull/59

hoechenberger replied:
Yep. We can / should discuss introducing this across the public API. Sometime later. ;)

hoechenberger replied:
In my personal projects, I've been using this for a long time, see e.g. https://github.com/hoechenberger/questplus/blob/master/questplus/qp.py#L10-L20

jasmainak replied:
I think it's a nice trick, specially for private functions because we often mess up when passing arguments around

jasmainak replied:
For public API, I wouldn't like to be forced to use keyword arguments if I was just passing one or two parameters. Maybe using the * after first 2 or 3 params is fine.

jasmainak commented on mne_bids/read.py:
```diff
@@ -328,7 +328,117 @@ def _handle_channels_reading(channels_fname, bids_fname, raw):
+def make_bids_fname(bids_basename, bids_root=None, kind=None, extension=None,
+                    verbose=False):
+    """
+    Construct the filename of a BIDS data file.
```
I think PEP8 says your L395 and L394 should be in a single line

jasmainak commented on mne_bids/read.py:
```diff
@@ -328,7 +328,117 @@ def _handle_channels_reading(channels_fname, bids_fname, raw):
+    return bids_fname
+
+
+def make_bids_fname(bids_basename, bids_root=None, kind=None, extension=None,
```
should be private?

jasmainak commented on mne_bids/read.py:
```diff
@@ -328,7 +328,117 @@ def _handle_channels_reading(channels_fname, bids_fname, raw):
+    data_dir = make_bids_folders(subject=sub, session=ses, kind=kind,
+                                 make_dir=False)
+
+    if op.isdir(op.join(bids_root, data_dir, f'{bids_basename}_{kind}')):
+        bti_dir = op.join(bids_root, data_dir, f'{bids_basename}_{kind}')
+        logger.info(f'Assuming BTi data in {bti_dir}')
+        bids_fname = f'{bti_dir}.pdf'
```
```suggestion
     bti_dir = op.join(bids_root, data_dir, f'{bids_basename}_{kind}')
     if op.isdir(bti_dir):
        logger.info(f'Assuming BTi data in {bti_dir}')
        bids_fname = f'{bti_dir}.pdf'
```

hoechenberger commented:
## Motivation
In `master`, `read_raw_bids()` expects users to pass a `bids_fname`, which consists of the **BIDS basename**, the **recording modality**, and the **filename extension**. Currently, it has to be constructed manually:
```python
bids_root = '/bids'
bids_basename = make_bids_basename(subject='foo', session='bar', task='baz')
kind = 'meg'
ext = '.fif'
fname = f'{bids_basename}_{kind}{ext}'
raw = read_raw_bids(bids_fname=bids_fname, bids_root=bids_root)
```

This comes with the following issues:
- The approach is **not portable** in the sense that the filename extension is hardcoded. Conversion to a different format, e.g. from FIFF to BrainVision for EEG, would break scripts the use `mne-bids` to process the data.
- The **API is inconsistent** when compared with `write_raw_bids()`, which requires a `bids_basename` instead of a `bids_fname` (and generates extension and `kind` from the provided raw data).

## Idea & Initial implementation
So my idea was to alter the `read_raw_bids()` signature from:
```python
 def read_raw_bids(bids_fname, bids_root, extra_params=None, 
                   verbose=True): 
```

to

```python
 def read_raw_bids(bids_basename, bids_root, kind, extra_params=None, 
                   verbose=True): 
```

Note that `bids_fname` is gone, while the two new parameters `bids_basename` and `kind` were introduced.

But to construct to complete filename, we also need a an **extension**. Because I wanted to eliminate the need to specify a hardcoded extension, I introduced a function `make_bids_fname()`:

```python
def make_bids_fname(bids_basename, bids_root, kind, extension=None,
                     verbose=False):
```
_As @jasmainak pointed out in a review, this function should probably be private._

This function first extracts the `sub` and `ses` entities from `bids_basename`, and then passes these values together with `kind` to `_get_bids_fname_from_filesystem()`. What this function, in turn, does it essentially checks if any data files can be found for the specified values, and it raises an exception if that's not the case or if there is ambiguity (i.e., multiple data files for this recording; in this case, an exception is raised suggesting the user run the `bids-validator` on their dataset).

After that, I realized I also had to update `get_matched_empty_room()` and `get_head_mri_trans()`:

old:
```python
def get_matched_empty_room(bids_fname, bids_root):
    ...

def get_head_mri_trans(bids_fname, bids_root):
    ...
```

new:
```python
def get_matched_empty_room(bids_basename, bids_root):
    ...

def get_head_mri_trans(bids_basename, bids_root):
    ...
```

## Result
Now you could do:

```python
bids_root = '/bids'
bids_basename = make_bids_basename(subject='foo', session='bar', task='baz')

# Read
kind = 'meg'
raw = read_raw_bids(bids_basename=bids_basename, bids_root=bids_root, kind=kind)

# Write
write_raw_bids(raw=raw, bids_basename=bids_basename, bids_root=bids_root)
```

Just to remind you, the equivalent code with `master` looks like this:
```python
bids_root = '/bids'
bids_basename = make_bids_basename(subject='foo', session='bar', task='baz')

# Read
kind = 'meg'
ext = '.fif'
fname = f'{bids_basename}_{kind}{ext}'
raw = read_raw_bids(bids_fname=bids_fname, bids_root=bids_root)

# Write
write_raw_bids(raw=raw, bids_basename=bids_basename, bids_root=bids_root)
```

Advantages:
- Less code to write
- better consistency between `read_raw_bids()` and `write_raw_bids()` signatures

Disadvantages:
- filename extension needs to be inferred, but hopefully this should never be a problem with BIDS-compliant datasets (?)

## How it blew up
For whatever reason, I then expanded the scope of the `make_bids_fname()` function to allow construction of a filename that is _not_ necessarily present in the current dataset, by explicitly passing an `extension`. Not sure why I did that anymore, and this feature can and should probably be removed.

Next I thought, if we're smart enough to infer the **extension**, why not be even smarter and infer the ** recording kind** too? Because if only one  kind is present in the dataset, it's peace of cake to extract that… So `_infer_kind()` and its auxiliary function _get_kinds_for_sub()` were born, which retrieves the available recoding kinds for a specific subject and session, and is invoked when you don't provide a `kind` to `read_raw_bids()`. So you would be able to do something like:
```python
bids_root = '/bids'
bids_basename = make_bids_basename(subject='foo', session='bar', task='baz')
raw = read_raw_bids(bids_basename=bids_basename, bids_root=bids_root)
```
and it would do The Right Thing and read the correct data if only _one_ recording kind is present , or raise an exception otherwise.

As you can see, complexity kept increasing… and now I also had to deal with situations where we'd have to infer both, recoding kind _and_ filename extension. All a bit confusing, at least _too_ confusing for my taste for _one_ PR!

## How to proceed
I suggest I strip down this PR in the following way:

- We keep the signature changes of `read_raw_bids()`, `get_matched_empty_room()`, `get_head_mri_trans()`, as this makes the API more consistent.
- `kind` becomes a required parameter of `read_raw_bids()`, so the `kind` inference can be removed
    - this will allow me to drop `_infer_kind()` and `_get_kinds_for_sub()`
- Make `make_bids_fname()` a private function and remove possibility to construct arbitrary filenames that don't exist on disk

Result, from a user's perspective, would then be:
```python
bids_root = '/bids'
bids_basename = make_bids_basename(subject='foo', session='bar', task='baz')
kind = 'meg'
raw = read_raw_bids(bids_basename=bids_basename, bids_root=bids_root, kind=kind)
```

### Other use cases
@agramfort said he'd like to be able to pass filenames directly to `read_raw_bids()`, as is currently the case with `master`. Within my proposed framework, we'd have two options to do this:

- `read_raw_bids()` could accept `bids_fname` as an optional kwarg; if passed, it will be used to load a file directly. This would also ensure backward-compatibility with existing code. We'd have to raise an exception if a user combines this kwarg with the new ones I proposed (`bids_basename` and `kind`). Seems a little awkward to me.
- We introduce a function that splits a filename into the `basename` and `kind`: 
    ```python
    def fname_to_basename_and_kind(fname):
        fname = pathlib.Path(fname)
        # FIXME use some regex foo instead…
        kind = fname.stem.split('_')[-1]
        basename = fname.stem[:-len(kind) - 1]
        return basename, kind
    ```

    This would allow one to do:
    ```python
    fname = '/foo/bar/sub-01/ses-01/sub-01_ses-01_task-foo_meg.fif'
    bids_basename, kind = fname_to_basename_and_kind(fname)
    read_raw_bids(bids_basename=bids_basename, kind=kind, bids_root='foo')
    ```

jasmainak commented:
> this will allow me to drop _infer_kind() and _get_kinds_for_sub()

I'm wondering why this is so complicated? You need to check if there is the directory `bids_root/sub-<subject_label>/meg` or `bids_root/sub-<subject_label>/eeg` or `bids_root/sub-<subject_label>/ieeg`. That's it. If two of them exist, throw an error asking user to provide it. To me, having `kind` as optional is the most consistent with the writing function API. It's not so hard to provide on the user end, but it's not so hard to code up either?

> We introduce a function that splits a filename into the basename and kind

sounds fair to me

jasmainak commented:
> if passed, it will be used to load a file directly. This would also ensure backward-compatibility with existing code.

I think we should do a deprecation cycle here. There shouldn't be 2 ways to do the same thing. Zen of Python :)

agramfort commented:
I agree that kind='auto' should be the default as @jasmainak suggested

regarding allowing passing full fname I would allow that bids_basename
accepts a file name with extension. But this can be left for another PR
after confirming that this is a necessary thing to support.

hoechenberger commented:
@jasmainak 
> I'm wondering why this is so complicated? You need to check if there is the directory `bids_root/sub-<subject_label>/meg` or `bids_root/sub-<subject_label>/eeg` or `bids_root/sub-<subject_label>/ieeg`. That's it.

No, you also have to account for the session :) This is all solved by my `_infer_kind()` function, which calls `_get_kinds_for_sub()`. Not much code, really. And now, looking at it again, I guess you're right: it's not that complicated.

> To me, having `kind` as optional is the most consistent with the writing function API.

Actually, to be consistent, we should also add `kind=None` to `write_raw_bids()`, because currently it would write multiple kinds if they are present in the input data, right?
But that's something for another PR :)

> I think we should do a deprecation cycle here. There shouldn't be 2 ways to do the same thing. Zen of Python :)

Yeah I do know the Zen but really think it shouldn't be too much of a dogma. In this particular case, however, I agree with you: we should have only one way to do this. :)

So my updated proposal to move forward is now:

In this PR:
- `make_bids_fname()` becomes a private function
- I address your [comment](https://github.com/mne-tools/mne-bids/pull/410/files#r426173419) re `kind` inference
- I add tests for `kind=None`

In a folloowup PR:
- Introduce a function to convert a filename to `bids_basename` and `kind`, to address @agramfort's use case

agramfort commented:
no objection but I will really have the feeling of the progress but noticing
a simplification of the code of examples

>

hoechenberger commented:
> no objection but I will really have the feeling of the progress but noticing a simplification of the code of examples

Could you please rephrase that? I'm not sure what you meant to say :) If you're asking for example code, I included some in my longer post from this afternoon:

If we proceed as planned, user code will be simplified from:

```python
bids_basename = make_bids_basename(subject='foo', session='bar', task='baz')
fname = f'{bids_basename}_meg.fif'  # Notice how we're hardcoding kind and extension here
raw = read_raw_bids(bids_fname=bids_fname, bids_root=bids_root)
```
to
```python
bids_basename = make_bids_basename(subject='foo', session='bar', task='baz')
raw = read_raw_bids(bids_basename=bids_basename, bids_root=bids_root)
```
for "single-kind" experiments.

hoechenberger commented:
Damn, looks like a screwed up a rebase somewhere… ugh. Will try to fix this mess.

hoechenberger commented:
Ok got it, my IDE apparently wouldn't properly pull in upstream/master, so I kept rebasing onto an outdated master branch…

hoechenberger commented:
@agramfort @jasmainak 
While updating the examples I realized that `get_matched_emptyroom()` currently returns a filename, so the current use is sth like:
```python
best_er_fname = get_matched_empty_room(bids_basename=bids_basename,
                                       bids_root= bids_root)
er_raw = read_raw_bids(best_er_fname, bids_path)
```
Obviously, this is in contrast to how I have refactored `read_raw_bids()`. So now the question is: what should `get_matched_empty_room()` return instead? The emptyroom `session`? Or the complete `basename`?

```python
# get_matched_empty_room() returns session
session = get_matched_empty_room(bids_basename=bids_basename, bids_root=bids_root)
basename  = make_bids_basename(..., session=session)
er_raw = read_raw_bids(bids_basename= basename, bids_root=bids_root)

# get_matched_empty_room() returns basename
basename = get_matched_empty_room(bids_basename=bids_basename, bids_root=bids_root)
er_raw = read_raw_bids(bids_basename= basename, bids_root=bids_root)
```
I personally like the latter much more…

hoechenberger commented:
Oh the joy of circular imports 🙈

hoechenberger commented:
Ok I think I'm done here, except for a changelog entry.

But now that I've changed so much, I'm wondering how to actually implement a proper deprecation cycle…

hoechenberger commented:
@sappelhoff @jasmainak 
If you think we could do this without a deprecation cycle, I can write a changelog entry and we're done… otherwise, things are going to be a little complicated

sappelhoff commented:
> I'm wondering how to actually implement a proper deprecation cycle…

good thing we agreed to not do that :)
(can't find the comment right now though ) 

> If you think we could do this without a deprecation cycle

yes, I still think so, trusting that (a) few people use this, and those that do are here with us on GitHub anyways and (b) if we break it for somebody we'll at least hear something from them in our issues.

hoechenberger commented:
> good thing we agreed to not do that :)

😅



> yes, I still think so, trusting that (a) few people use this (currently, hopefully more in the future), and those that do are here with us on GitHub anyways and (b) if we break it for somebody we'll at least hear something from them in our issues.

👍

hoechenberger commented:
Thanks. Do I need to rebase & squash or would you rather do this through GitHub while merging?

agramfort commented:
do it with github

i just want @jasmainak to confirm it's ok for him

jasmainak commented on mne_bids/read.py:
```diff
@@ -340,58 +341,184 @@ def _handle_channels_reading(channels_fname, bids_fname, raw):
+            raise RuntimeError(msg)
+
+        # FIXME This will break e.g. with FIFF data split across multiple
+        # FIXME files.
```
we normally use XXX but FIXME is also fine I guess

jasmainak commented on mne_bids/read.py:
```diff
@@ -551,10 +683,27 @@ def get_matched_empty_room(bids_fname, bids_root):
+    if best_er_fname is None:
+        er_basename = None
+    else:
+        from mne_bids import make_bids_basename  # Avoid circular import.
```
okay this needs to be fixed in another PR

jasmainak commented:
The PR looked really neat, loved it! Thanks @hoechenberger :)

jasmainak commented:
I commented on a few things but I don't think they were critical for this PR to be merged, they can be fixed elsewhere. Also one thing that might help is to clarify consistently everywhere in the docs what `bids_basename` means. It's really a mne-bids specific concept, not someone familiar with BIDS would understand.
